debug_bundle: Added label_select option

michael-redpanda opened the pull request:
Added ability to select labels via rpk debug bundle by providing a key and value for labels.

Using this will pass the key value pairs as key=value to the --label-selector option of rpk debug bundle.

Fixes: CORE-7859

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

BenPope commented on src/v/debug_bundle/debug_bundle_service.cc:
```diff
@@ -677,6 +713,26 @@ result<std::vector<ss::sstring>> service::build_rpk_arguments(
+      params.label_selector.has_value()
+      && !params.label_selector.value().empty()) {
+        rv.emplace_back(k8s_label_selector);
+        chunked_vector<ss::sstring> selectors;
```
nit: It might be worth putting a reserve on here, although if you had a:
```c++
    friend std::ostream&
    operator<<(std::ostream& os, const label_selection& l) {
        fmt::print(os, "{}={}", l.key, l.value);
        return os;
    }
```
You could probably avoid this container.

BenPope reviewed (commented):
LGTM, but seems to be missing the swagger

IoannisRP commented on src/v/debug_bundle/debug_bundle_service.cc:
```diff
@@ -109,6 +110,41 @@ bool is_valid_k8s_namespace(std::string_view ns) {
     return !ns.empty() && ns.size() <= max_ns_length && is_valid_rfc1123(ns);
 }
 
+bool is_valid_k8s_label_name_or_value(std::string_view name) {
+    constexpr auto max_label_name_size = 63;
+    constexpr std::string_view label_name_pattern
+      = R"(^([a-zA-Z0-9]([a-zA-Z0-9-_.]{0,61}[a-zA-Z0-9])?))";
+
+    return !name.empty() && name.size() <= max_label_name_size
+           && RE2::FullMatch(name, RE2(label_name_pattern));
+}
+
+// This function doesn't fully check that the label prefix is a 100% valid K8S
+// label prefix but at least verifies that the contained characters are valid.
+// rpk will validate using the golang labels.go library
+bool is_likely_valid_k8s_label_prefix(std::string_view prefix) {
+    constexpr auto prefix_max_length = 253;
+    constexpr std::string_view label_name_pattern
+      = R"(^([a-zA-Z0-9]([a-zA-Z0-9-_.]{0,251}[a-zA-Z0-9])?))";
+    return !prefix.empty() && prefix.size() <= prefix_max_length
+           && RE2::FullMatch(prefix, RE2(label_name_pattern));
```
Could the bodies of these two function be merged into a single one?

They seem to be parameterize-able on the length of the max_length. I assume that the benefit of having constexpr values is negligible compared to the runtime of regex.

BenPope replied:
If the refactoring is worth doing, potentially merge and and/alias `is_valid_rfc1123` into `is_valid_k8s_namespace`, and make the `RE2`s `static const` so that they are compiled once.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--7b0530b1-14c4-489a-821d-d541932af9a5-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/56119#01927215-0f8a-48cf-83df-ab6d0c9bf9a8
<!--/body-->

michael-redpanda commented:
Force push [`45d1dc3`](https://github.com/redpanda-data/redpanda/compare/dec162dbece9ac39d285e1a9e050eb1336c61415..45d1dc3a0b70eddc093c9ce092a87dfbe2a32a11):

* Updated swagger
* Combined label checking function
* Fixed up label selector parameter printing

BenPope commented on src/v/redpanda/admin/api-doc/debug_bundle.json:
```diff
@@ -81,6 +81,24 @@
                   "type": "string"
+                },
+                "label_selector": {
+                  "type": "list",
```
```suggestion
                  "type": "array",

BenPope replied:
Everything else LGTM

michael-redpanda commented:
Force push [`76cb8b7`](https://github.com/redpanda-data/redpanda/compare/45d1dc3a0b70eddc093c9ce092a87dfbe2a32a11..76cb8b75b72fdecf52125f67f0ea72020d2eec85):

* Swapped `list` for `array` in swagger

dotnwat commented on src/v/debug_bundle/types.h:
```diff
@@ -88,6 +88,18 @@ struct partition_selection {
 std::ostream& operator<<(std::ostream& o, const partition_selection& p);
 
 using k8s_namespace = named_type<ss::sstring, struct k8s_namespace_tag>;
+using k8s_label_key = named_type<ss::sstring, struct k8s_label_key_tag>;
+using k8s_label_value = named_type<ss::sstring, struct k8s_label_value_tag>;
```
why do we need anything `k8s` specific in the tree? can't we provide a mechanism to pass arbitrary key-value pairs?

dotnwat replied:
but why does redpanda need to validate that?

michael-redpanda replied:
I'm validating length and character types.  I'm not doing a full validation and allowing rpk to do it.  The length/character type check is present to ensure no 'weird' characters are present and the length check is to ensure we aren't passing a massive string with null characters in it.

This does remind me to add in some additional checks for the other strings we handle (password/username etc).

dotnwat replied:
> The length/character type check is present to ensure no 'weird' characters are present and the length check is to ensure we aren't passing a massive string with null characters in it.

sounds like these are generic checks, not specific to k8s? i'm all for validation, but k8s specific?

michael-redpanda replied:
Actually in the process of removing them.  Gave it a thought and remembered the validate_no_control function.  Going to implement using that against all provided strings and remove these checks.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/56299-->
<!--c090b340-0bee-4d5d-aacf-2872a26f6886-->
<!--title-->
## Retry command for [Build#56299](https://buildkite.com/redpanda/redpanda/builds/56299)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage@{"cleanup_policy":"delete"}
```
<!--/body-->

Deflaimun commented on src/v/redpanda/admin/api-doc/debug_bundle.json:
```diff
@@ -81,6 +81,24 @@
                 "namespace": {
                   "type": "string"
+                },
+                "label_selector": {
```
no description?

michael-redpanda commented:
Force push [`1840b4c`](https://github.com/redpanda-data/redpanda/compare/eebbac064487cb97052c3cb78cd716a6b48de856..1840b4c6f2c5a0d05e3eaa5601e651e27a41633d):

* Added in control character checks for strings provided via json parameters
* Removed k8s checks in favor of just the control character checks
* Addressed review comments

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/56317-->
<!--81e15525-4c44-4e90-8b46-029bf3f12a66-->
<!--title-->
## Retry command for [Build#56317](https://buildkite.com/redpanda/redpanda/builds/56317)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage@{"cleanup_policy":"delete"}
```
<!--/body-->

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--81dd331f-3f92-4eed-af87-ca1ddf54e6eb-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/56317#019279bc-d771-4c1e-9334-f1026e67b66d:
```
"rptest.tests.cluster_config_test.ClusterConfigTest.test_valid_settings"
```
<!--/body-->

michael-redpanda commented:
/ci-repeat 1
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage@{"cleanup_policy":"delete"}
tests/rptest/tests/cluster_config_test.py::ClusterConfigTest.test_valid_settings

michael-redpanda commented on src/v/debug_bundle/json.h:
```diff
@@ -83,7 +84,14 @@ debug_bundle::result<T> from_json(const json::Value& v) {
-            return T{as_string_view(v)};
+            auto vv = as_string_view(v);
+            try {
+                validate_no_control(vv);
```
Would that error on control characters?  I think they're permitted by the JSON spec

BenPope replied:
> Would that error on control characters? I think they're permitted by the JSON spec

You are correct.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/56342-->
<!--47fb6cc5-79fa-4427-b616-3fe1b0bb0776-->
<!--title-->
## Retry command for [Build#56342](https://buildkite.com/redpanda/redpanda/builds/56342)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_storage_timing_stress_test.py::CloudStorageTimingStressTest.test_cloud_storage@{"cleanup_policy":"delete"}
```
<!--/body-->

michael-redpanda commented:
CI Failure:
* [CI failure (Incomplete checks) in CloudStorageTimingStressTest.test_cloud_storage](https://redpandadata.atlassian.net/browse/CORE-7002)

vbotbuildovich commented:
Oops! Something went wrong.

[Workflow run logs][1].

[1]: https://github.com/redpanda-data/redpanda/actions/runs/11293872297

michael-redpanda commented:
/publish-to-install-pack

vbotbuildovich commented:
Oops! Something went wrong.

[Workflow run logs][1].

[1]: https://github.com/redpanda-data/redpanda/actions/runs/11293911805
